File: Generate-Parentheses.cpp
class Solution {
public:
    void solve(int openP, int closeP, string s, int n, vector<string>& res) {
        if (openP == closeP && openP + closeP == n * 2) {
            res.push_back(s);
            return;
        }

        if (openP < n) {
            solve(openP + 1, closeP, s + \(\, n, res);
        }

        if (closeP < openP) {
            solve(openP, closeP + 1, s + \)\, n, res);
        }
    }
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        solve(0, 0, \\, n, res);
        return res;        
    }
};

File: Generate-Parentheses.py
class Solution:


        
    def generateParenthesis(self, n: int) -> List[str]:

        res =[]
        temp = \\

        def solve(openP, closeP, n, res, temp):

            if (openP+closeP == n*2):
                res.append(temp)
                return
            
            if(openP<n):
                solve(openP+1, closeP, n, res, temp+\(\)

            if(closeP<openP):
                solve(openP, closeP+1, n, res, temp+\)\)

        solve(0, 0, n, res, temp)
        return res


        